clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java

```diff
@@ -177,12 +177,15 @@ public static PartitionInfo toPartitionInfo(PartitionMetadata metadata, Map<Inte
     }
 
     private static Node[] convertToNodeArray(List<Integer> replicaIds, Map<Integer, Node> nodesById) {
-        return replicaIds.stream().map(replicaId -> {
+        Node[] nodes = new Node[replicaIds.size()];
+        for (int i = 0; i < replicaIds.size(); i++) {
+            Integer replicaId = replicaIds.get(i);
```

Comment:
nit:

```suggestion
            int replicaId = replicaIds.get(i);
```

Reply:
We could unbox this to `int`, but it would be boxed back to `Integer` on the next line as argument to `nodesById.get(replicaId)` for `Map<Integer, Node> nodesById`. For `replicaId` < 128 (unless one increases/decreases `java.lang.Integer.IntegerCache.high` system property) this would be cached by `java.lang.Integer`'s internal cache, otherwise it would allocate a new `Integer`.

I'm inclined to leave it as `Integer replicaId = replicaIds.get(i);` to avoid that roundtrip.

Reply:
Sounds good.